diff --git a/drivers/usb/core/Kconfig b/drivers/usb/core/Kconfig
index 351ede4b5..660d5728c 100644
--- a/drivers/usb/core/Kconfig
+++ b/drivers/usb/core/Kconfig
@@ -108,7 +108,7 @@ config USB_LEDS_TRIGGER_USBPORT
 config USB_AUTOSUSPEND_DELAY
 	int "Default autosuspend delay"
 	depends on USB
-	default 2
+	default 1
 	help
 	  The default autosuspend delay in seconds.  Can be overridden
 	  with the usbcore.autosuspend command line or module parameter.
diff --git a/kernel/power/Kconfig b/kernel/power/Kconfig
index a12779650..08f3a9963 100644
--- a/kernel/power/Kconfig
+++ b/kernel/power/Kconfig
@@ -291,6 +291,7 @@ config PM_GENERIC_DOMAINS
 config WQ_POWER_EFFICIENT_DEFAULT
 	bool "Enable workqueue power-efficient mode by default"
 	depends on PM
+	default y
 	help
 	  Per-cpu workqueues are generally preferred because they show
 	  better performance thanks to cache locality; unfortunately,
diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 3d90d42a7..5d513ba58 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -49,7 +49,7 @@ config STACKTRACE_BUILD_ID
 config CONSOLE_LOGLEVEL_DEFAULT
 	int "Default console loglevel (1-15)"
 	range 1 15
-	default "7"
+	default "4"
 	help
 	  Default loglevel to determine what will be printed on the console.
 
